java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -959,53 +964,68 @@ class DatabaseImpl(
         schema: Schema,
         db: SQLiteDatabase,
         counters: Counters? = null
-    ): TypeId = schemaMutex.withLock {
-        schemaTypeMap[schema.hash]?.let {
-            counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_HIT)
-            return@withLock it
-        }
-        counters?.increment(DatabaseCounters.ENTITY_SCHEMA_CACHE_MISS)
-
-        return db.transaction {
-            counters?.increment(DatabaseCounters.INSERT_ENTITY_TYPE_ID)
+    ): TypeId = db.transaction {
+        var cacheHit = false
+        val result = schemaTypeMap.updateAndGet { currentMap ->
+            currentMap[schema.hash]?.let {
+                cacheHit = true
+                return@updateAndGet currentMap
+            }
+            cacheHit = false
             val content = ContentValues().apply {
                 put("name", schema.hash)
                 put("is_primitive", false)
             }
-            val schemaTypeId = insertOrThrow(TABLE_TYPES, null, content)
+            val id = insertWithOnConflict(
+                TABLE_TYPES,
+                null,
+                content,
+                SQLiteDatabase.CONFLICT_IGNORE
+            )
+            if (id >= 0) {
+                currentMap + (schema.hash to id)
+            } else {
+                cacheHit = true
+                currentMap
+            }
+        }.getOrDefault(schema.hash, 3)
```

Comment:
what is 3? isn't it guaranteed that schema.hash is in the map?

Reply:
Yeah, 3 was just a non-primitive/collection type value as a default. Probably not a smart idea though.